062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,254 @@
+| STRIMZI_CRUISE_CONTROL_SECRET_NAME     | String  | ""      | Secret name containing the truststore                         |
+| STRIMZI_CRUISE_CONTROL_API_SECRET_NAME | String  | ""      | Secret name containing the API credentials                    |
+
+For the standalone deployment, the Topic Operator expects the following secrets:
+
+- A secret with `cruise-control.p12` key containing the truststore in PKCS12 format, and `cruise-control.password` key containing the truststore password.
+- A secret with `cruise-control.apiToAdminUsername` key containing the API username with `ADMIN` role, and `cruise-control.apiToAdminPassword` key containing the user password.
+
+Cruise Control's topic configuration endpoint can be used to request a replication factor change.
+Once accepted, the replication factor change is put into a task queue and executed asynchronously.
+A task is first created in `Active` state, then moved to `InExecution`, and finally to `Completed` or `CompletedWithError`.
+Only a single task can be `InExecution` at any given time, which includes both topic configuration and rebalance tasks.
+
+When Cruise Control integration is enabled, a `ReplicasChangeManager` is created to manage the communication with the REST API.
+The `ReplicasChangeManager` stores all pending and ongoing replication factor changes in an in-memory cache used to coordinate status updates and periodically check tasks.
+
+<figure>
+    <img src="images/062-topic-replication-factor-change-arch.png" height=450>
+    <figcaption><small>
+        Fig 2. High level architecture.
+    </small></figcaption>
+</figure>
+
+<br/>
+<br/>
+
+Cruise Control allows to scale down the replication factor under the `min.insync.replicas` value, and this can cause an availability issue for producers with `acks=all`.
+The Topic Operator logs a warning about that, but continues with the change because the `Kafka Roller` ignores topics with RF < minISR, as they don't event show up as under replicated.
+The target replication factor should also be less than or equal to the number of available brokers, but this is enforced directly by Cruise Control.
+
+List of the most common Cruise Control exceptions:
+
+- `NotEnoughValidWindowsException` can be raised if the cluster workload model is not ready when a replication factor change is requested.
+  This may happen when Cruise Control has just started and does not have enough samples, or when there is some issue in producing and consuming metrics.
+  In this case, the Topic Operator replaces Cruise Control's error message with a custom warning message.
+- `OngoingExecutionException` can be raised if there is already a task that is being executed.
+  In this case, the Topic Operator replaces Cruise Control's error message with a custom warning message.
+- `OptimizationFailureException` can be raised if the target replication factor is smaller than the number of available brokers, or hard goals are violated.
```

Comment:
`hard goals are violated` ... is this referring to rebalance, or there are "hard goals" to be met on replication factor change as well?

Reply:
The underline mechanism is the same, so goals are also used for RF changes.